### DISCUSSION TITLE
Cannot parse query parameter list from request object

### DISCUSSION THREAD

**Discussion by bsandeepan** [fastapi/fastapi#6550] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import uvicorn
from fastapi import FastAPI, Query, Request
from typing import Optional, List
from pydantic import BaseModel

app = FastAPI(
    title="Test API",
    description="This app does nothing",
)

class UsersGetRequest(BaseModel):
    """ Parse and validate incoming request query parameters types and values (basic). """
    
    page_size: int = Query(default=100, le=200, title="page_size")
    name: Optional[str] = Query(default=None, min_length=2, max_length=256, title="name", example="sandeepan b")
    get_user_count: Optional[bool] = Query(default=False, title="get_user_count")
    address: Optional[int] = Query(default=None, ge=0, le=2, title="address")
    locations: Optional[List[str]] = Query(default=None, title="locations")


@app.get("/")
def send_user(
    request: Request,
    locations: Optional[List[str]] = Query(default=None, title="locations")
):
    incoming_data = UsersGetRequest.parse_obj(request.query_params)
    return incoming_data.dict()

if __name__ == "__main__":
    uvicorn.run(app, host="localhost", port=9999)
```

### Description

We have a microservice following CQRS pattern. We usually parse query parameters or bodies using DTOs to keep code as much clean as possible. Now, I am using such DTO to parse query parameters for an endpoint which is working fine in general. However, we require some params to be list of values now. Trying to implement this is giving me problem. If I do it in the path operation function, then it works fine. However, I have about 20+ query parameters and this list will only grow in future. Is there any better way to handle list of inputs in FastAPI/Pydantic other than defining parameters in the function definition?

### Operating System

Linux

### FastAPI Version

0.65.2

### Python Version

3.7.10

### Additional Context

Note: 
- Earlier, I faced similar issue when trying to handle list of files in a POST method. Had to define the parameter in function definition. Are these linked?
- I think my issue is similar to this one: https://github.com/tiangolo/fastapi/issues/2869 but this has no proper answer.

**Comment by frankie567** [comment]
A more "FastAPI way" to do this is to implement your query parameter parsing in a separate class dependency (see https://fastapi.tiangolo.com/tutorial/dependencies/classes-as-dependencies/#classes-as-dependencies_1):

```py
import uvicorn
from typing import List, Optional

from fastapi import FastAPI, Depends, Query

app = FastAPI(
    title="Test API",
    description="This app does nothing",
)


class UsersGetRequest:
    def __init__(
        self,
        page_size: int = Query(default=100, le=200, title="page_size"),
        name: Optional[str] = Query(default=None, min_length=2, max_length=256, title="name", example="sandeepan b"),
        get_user_count: Optional[bool] = Query(default=False, title="get_user_count"),
        address: Optional[int] = Query(default=None, ge=0, le=2, title="address"),
        locations: Optional[List[str]] = Query(default=None, title="locations"),
    ):
        self.page_size = page_size
        self.name = name
        self.get_user_count = get_user_count
        self.address = address
        self.locations = locations


@app.get("/")
def send_user(user_get_request: UsersGetRequest = Depends(UsersGetRequest)):
    return user_get_request


if __name__ == "__main__":
    uvicorn.run(app, host="localhost", port=9999)
```

This way, you purely embrace the dependency injection pattern. The automatic documentation is fully working, reporting every parameters. It's also reusable across several endpoints.

**Reply by JoshYuJump** [reply]
I have tried the solution, `locations` is query param but it's not array in swagger(OpenAPI).

**Comment by bsandeepan** [comment]
@frankie567 thanks for nudging me in the right direction! I checked the documentation and yes, it worked, and more!
I followed the example in the page and spun up similar flow using pydantic models, as that is what I need the most, for running additional validations using `@validator`.

```python3
import pycountry
import uvicorn
from fastapi import FastAPI, Depends, Query
from typing import Optional, List
from pydantic import BaseModel, validator

app = FastAPI(
    title="Test API",
    description="This app does nothing",
)

class UsersGetRequest(BaseModel):
    """ Parse and validate incoming request query parameters types and values (basic). """

    page_size: int
    name: Optional[str]
    get_user_count: Optional[bool]
    address: Optional[int]
    locations: Optional[List[str]]

    def __init__(self,
        page_size: int = Query(default=100, le=200, title="page_size"),
        name: Optional[str] = Query(default=None, min_length=2, max_length=256, title="name", example="sandeepan b"),
        get_user_count: Optional[bool] = Query(default=False, title="get_user_count"),
        address: Optional[int] = Query(default=None, ge=0, le=2, title="address"),
        locations: Optional[List[str]] = Query(default=None, title="locations"),
    ):
        super().__init__(**{key: value for key, value in locals().items() if key != 'self'})
        # super().__init__(
        #    page_size=page_size,
        #    name=name,
        #    get_user_count=get_user_count,
        #    address=address,
        #    locations=locations
        # )

    @validator('locations', each_item=True)
    def check_locations(cls, v: str):
        assert pycountry.countries.get(name=v) is not None, "Incorrect Country Location"
        return v


@app.get("/")
def send_user(incoming_data: UsersGetRequest = Depends(UsersGetRequest)):
    return incoming_data

if __name__ == "__main__":
    uvicorn.run(app, host="localhost", port=9999)
```

Though it has more code, it is necessary as I am trying to keep any additional model validations within model class itself. 
I also shortened the manual declaration in `super().__init__()` now.

If anyone knows a better way to do this, please guide me. Thanks!

**Comment by bsandeepan** [comment]
I am closing the issue  as the problem I had is solved. However, I still request further improvements on this.

**Comment by dhessdiversigen** [comment]
This is still broken, I'm using fastapi 0.104.0, and here's some simple example code:

```py
from fastapi import Depends, FastAPI, Query
from pydantic import BaseModel

app = FastAPI()


class Test(BaseModel):
    value: list[str] | None = Query(None)


@app.get("/v1")
def pydantic_model(test: Test = Depends(Test)):
    return {"value": test.value}


@app.get("/v2")
def non_pydantic_model(value: list[str] | None = Query(None)):
    return {"value": value}
```

Then hit `http://localhost:3005/v1?value=ihtopeqw` (or whatever port you use of course) and I just get back `{"value": null}`
Trade out v1 for v2, and it works fine.

As an additional note, on the swagger docs that are generated, v2 looks as you'd expect, but v1 is interpreting it as a request body:
![image](https://github.com/tiangolo/fastapi/assets/85644753/06966155-9311-45cc-95a1-d5a0033c826f)

**Reply by dhessdiversigen** [reply]
From my basic investigation, the issue lies when it analyzes the parameters `fastapi.dependencies.utils:analyze_param`
Due to how the signature works, when it's passed in. See the difference between the signature of the function, and the pydantic model:

```py
>>> inspect.signature(Test)
<Signature (*, value: list[str] | None = None) -> None>
>>> inspect.signature(non_pydantic_model)
<Signature (value: list[str] | None = Query(None))>
```

That default (`None` for the pydantic model, `Query(None)` for the func) is what's used for `value` in this function. 
This means the if check `elif isinstance(value, FieldInfo)` fails for the former, and works for the latter. So the former falls through to the `elif not field_annotation_is_scalar(annotation=type_annotation)` check later on, and `field_info` now gets set to `params.Body`


I don't know this stuff deep enough to know how to fix this though.

**Reply by GianlucaFicarelli** [reply]
To add more, I see that with the following code, `/v1` works when using `fastapi==0.104.1` and `pydantic==1.10.13`,
but it doesn't when using `fastapi==0.104.1` and `pydantic==2.5.1` (or any other pydantic v2 I tried):

```python
from typing import Annotated
from fastapi import Depends, FastAPI, Query
from pydantic import BaseModel

app = FastAPI()

class Test(BaseModel):
    value: Annotated[list[str] | None, Query()] = None

@app.get("/v1")
def pydantic_model(test: Annotated[Test, Depends(Test)]):
    return {"value": test.value}
```

Compared to the original code, value is defined as
```python
    value: Annotated[list[str] | None, Query()] = None
```
instead of
```python
    value: list[str] | None = Query(None)
```
although I expect them to behave the same.

**Reply by GianlucaFicarelli** [reply]
A workaround working with pydantic v2 is to use `dataclass` from `pydantic.dataclasses`:
```
from typing import Annotated
from fastapi import Depends, FastAPI, Query
from pydantic.dataclasses import dataclass

app = FastAPI()

@dataclass
class Test:
    value: Annotated[list[str] | None, Query()] = None

@app.get("/v1")
def pydantic_model(test: Annotated[Test, Depends(Test)]):
    return {"value": test.value}
```
It might not always be applicable, though.

**Reply by JoshYuJump** [reply]
It's a bug of Pydantic 
https://github.com/pydantic/pydantic/issues/7978

**Comment by YuriiMotov** [comment, marked as answer]
This works fine since FastAPI 0.115.0:

```python
import pycountry
from fastapi import FastAPI, Query
from typing import Optional, List
from pydantic import BaseModel, validator

app = FastAPI(
    title="Test API",
    description="This app does nothing",
)


class UsersGetRequest(BaseModel):
    """Parse and validate incoming request query parameters types and values (basic)."""

    page_size: int
    name: Optional[str]
    get_user_count: Optional[bool]
    address: Optional[int]
    locations: Optional[List[str]]

    @validator("locations", each_item=True)
    def check_locations(cls, v: str):
        assert pycountry.countries.get(name=v) is not None, "Incorrect Country Location"
        return v


@app.get("/")
def send_user(incoming_data: UsersGetRequest = Query()):
    return incoming_data

```
